### PR TITLE
lib: Introduce generic (page guarded) stack allocator library

### PR DESCRIPTION
**Pull request by mogasergiu**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A

### Description of changes

Implement a generic stack allocation wrapper over a given Unikraft
memory allocator. The library receives as arguments an allocator and,
if libukvmem is enabled, a virtual address space and an initial size.
The library works in three ways:
1. If libukvmem is disabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks.
2. If libukvmem is enabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks. The difference from 1. is that the virtual
address space and the initial size arguments are now available. The
two arguments will hint how much of the stack to premap whenever they
are allocated first, i.e. what is the initial stack size before
on-demand paging kicks in.
3. If libukvmem is enabled AND the configuration option of the library
integrating page guards is enabled, the allocator given as argument
will only be used once when allocating the initial allocator structure,
while the virtual address space and initial size arguments will be used
to create stack VMA's, specific to libukvmem.

With this addition, now libukboot is made aware of this stack allocator
and therefore a new configuration option is introduced to allow the
customization of the aforementioned initial size argument for when
libukvmem is enabled. Consequently, libukschedcoop's initialization
allows now for three additional allocator arguments for regular
stacks, for auxiliary stacks and for the TLS. This is done so that
libukboot can now initialize and pass to libukschedcoop separate
stack allocators, differentiated by the differently configured size,
alignment as well as the fact that, unlike regular stacks,
the auxiiliary stacks must always be fully mapped.

Also a quick basic commit sprinkling some error messages in the frame allocator :man_shrugging:.

### CODE REVIEW

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [inline lib/ukallocstack/stack.c]: Missed a spot, no trailing comma here.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [inline lib/ukallocstack/stack.c]: Note: just noticed I have lots of `CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS` instead of `CONFIG_LIBUKVMEM` comments.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [inline lib/ukallocstack/stack.c]: TODO: s/stackalloc/allocstack

File: lib/ukallocstack/include/uk/allocstack.h
Comment (michpappas) [inline lib/ukallocstack/include/uk/allocstack.h:53]: As this is the public API, please consider documenting to avoid confusion (e.g. what is the purpose of `initial_size`), incl. all provided ops and the constraints in their parameters. This is especially important for allocator-specific variations from the usual constraints, such as the alignment restrictions in `[posix_]memalign()`.

File: lib/ukallocstack/include/uk/allocstack.h
Comment (michpappas) [inline lib/ukallocstack/include/uk/allocstack.h]: Should be `#endif /* CONFIG_LIBUKVMEM */`

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: Should be `#endif /* CONFIG_LIBUKVMEM */`.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: ~~Also free(sa)~~ Sorry I didn't read this correctly.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: `unlikely`?

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c:20]: I would had called these `sa->sa` and `sa->a` instead of `sa->a` and `sa->og_a` mostly because I don't like the `og_a` vary much, but maybe that's a matter of personal taste, so feel free to ignore if you prefer the current convention.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: Heheh, at least you didn't use [`desired_size`](https://github.com/unikraft/unikraft/pull/1212/files#diff-628d5113b18ee6253ddbcc848991834d8493a6ad8684eea4e944be9252f73340L79) 🤤 this time. `alloc_size` could be a more elegant name.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: Perhaps it would be useful to add a comment about the purpose of that extra page.

File: lib/ukallocstack/Config.uk
Comment (michpappas) [inline lib/ukallocstack/Config.uk]: Not sure if the "wrapper" part is accurate here, given that if guards are enabled it is no longer just a wrapper. See also the comment below.

File: lib/ukallocstack/Config.uk
Comment (michpappas) [inline lib/ukallocstack/Config.uk:8]: So the main value of this library is the guards. And being able to use the library without vmem is just a convenience that allows the user to make a single call without conditionals.

Given that, I would probably reverse the logic and state here that without stack guards enabled this acts as a simple wrapper around the allocator. Similarly, in the description of the guards, I wouldn't describe it as an exception of the wrapper case (i.e. "instead of being a wrapper").

PS: There is also a school of thought that says that you should not describe the implementation at all, but only what the feature provides.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: You may consider making it a requirement that the user passes an aligned `initial_size`, so that you don't need to align up in every individual op.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c:78]: You may want to take into account the possibility that `size` < `initial_size`, or require that the user passes `size` >= `initial_size`.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: Same here, `UK_BUGON(rc)` if you choose to adopt the previous suggestion.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: Maybe it's a matter of personal taste but as far as semantics go `UK_BUGON(!vma)` would be more suitable here.

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: `unlikely`?

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: `unlikely`?

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c]: So on the one hand you make this check here, but on the other hand you delegate the `size` check to `stack_memalign()`. Maybe choose one scheme to be consistent?

File: lib/ukallocstack/stack.c
Comment (michpappas) [inline lib/ukallocstack/stack.c:173]: Shouldn't you adjust the return vaddr to the requested alignment?

**Review by michpappas — CHANGES_REQUESTED** [review]
@mogasergiu nice 😎  Some review comments below.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [reply lib/ukallocstack/stack.c:20]: I think I like `sa->a` and `sa->parent_a` :+1:

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [reply lib/ukallocstack/stack.c]: Regarding this and the other `unlikely` comments. I meant to align with the default `memalign`/`posix-memalign` checks which don't have these. But now, looking back, it's better to have them here in this case, since this lib is only meant to be used by our own kernel internal allocators :+1:

File: arch/Config.uk
Comment (michpappas) [inline arch/Config.uk:47]: I am posting this here as GitHub does not provide a way to comment on the commit message: 

`STACK_SIZE` should not had been under `arch/` in the first place. As discussed before, we should be moving everything not defined by the architecture away from `arch/`. Therefore this change does not improve something, both the current and the new place will be changed again shortly.

An additional problem is that this change propagates to an additional change in Kconfig. In principle Kconfig directives is something we should treat as a stable ABI and limit changes on them to the absolutely necessary, as such changes have an impact both on the catalog (they must be propagated to Kraftfiles) and potentially to systems maintained by adopters of Unikraft - that's even worse as we have no control over these. Of course under re-arch these are expected to change a lot, but still we should avoid them when they are not necessary.

Moreover, various stack sizes it makes more sense for the various stack sizes to be defined by the platform rather than the architecture. My take is that they should live in the upcoming `platform.h`. So I would propose that we drop this change, and move them to `platform.h` as soon as we have consensus on that file with the rest of the maintainers (I'll bring it up on the next call). Alternatively, instead of dropping - and assuming there will be consensus on `platform.h` - we can defer merging and update this commit to move them to that place. What do you think?

File: lib/ukvmem/Config.uk
Comment (michpappas) [inline lib/ukvmem/Config.uk:76]: Although these are not costly, similarly to timeouts there are no good default values. So I would be curious to know where the selected default come from, are 4 pages found to be good enough for most cases? Also, how common are accesses beyond 1 page?

File: lib/ukvmem/include/uk/vma_types.h
Comment (michpappas) [inline lib/ukvmem/include/uk/vma_types.h:127]: Same here. If there is a requirement for increasing the number of guard pages to >1, then it would be worth providing the reasoning in the commit message.

Moreover, should any of the tests that involve guards need to be updated? I did a quick run and they all pass, but didn't look at the implementation to determine whether they are still valid or they just produce false positives.

File: lib/ukvmem/Config.uk
Comment (michpappas) [inline lib/ukvmem/Config.uk]: For the stack base, is it meaningful to have more than 1 page? The only case I can think of is a case where a bug (benign or not) causes an overwrite beyond the stack base.

File: lib/ukvmem/include/uk/vma_types.h
Comment (michpappas) [inline lib/ukvmem/include/uk/vma_types.h:136]: I see what you did there 🎄

**Review by michpappas — CHANGES_REQUESTED** [review]
@mogasergiu a few more

File: lib/ukvmem/Config.uk
Comment (mogasergiu) [reply lib/ukvmem/Config.uk:76]: It's based on intuition after noticing that 2 pages are not enough for a Wordpress based application.

File: lib/ukvmem/include/uk/vma_types.h
Comment (mogasergiu) [reply lib/ukvmem/include/uk/vma_types.h:127]: > Moreover, should any of the tests that involve guards need to be updated?

They are still fine, the new guard pages also include the old guard page.

File: lib/ukvmem/Config.uk
Comment (mogasergiu) [reply lib/ukvmem/Config.uk]: > The only case I can think of is a case where a bug (benign or not) causes an overwrite beyond the stack base.

Yep, following a discussion with @skuenzer it was decided to make the guard pages end-to-end.

File: lib/ukvmem/include/uk/vma_types.h
Comment (mogasergiu) [reply lib/ukvmem/include/uk/vma_types.h:136]: World domination next :christmas_tree: :santa:

File: arch/Config.uk
Comment (michpappas) [reply arch/Config.uk:47]: **Update:** During an offline discussion with @skuenzer it was agreed that we merge this as is, given that migration involves additional code that existed previously in `arch`, and do a migration once all depending PRs are merged.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [reply lib/ukallocstack/stack.c]: hmm i think it's actually not needed anymore as I [removed](https://github.com/unikraft/unikraft/pull/1322/commits/3bf2bc5de8824ae55bd812ec011d9be27914a2c6#diff-af0f86d6a09f05c4a44e859d58dfe38ab95d442f01f60e26f92b345aae167f8fR124) this constraint.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [reply lib/ukallocstack/stack.c:78]: It's fine actually, that would mean the whole stack would be paged-in. I will document this.

File: lib/ukallocstack/stack.c
Comment (mogasergiu) [reply lib/ukallocstack/stack.c:173]: If using page guards then we allocate stack VMA's instead. So, the return address will always be page-aligned and the maximum allowed alignment is PAGE_SIZE because we cannot satisfy higher alignments and smaller than PAGE_SIZE alignments must be a divisor of PAGE_SIZE.

I wil add a comment

Comment (mogasergiu) [reply lib/ukallocstack/stack.c:173]: The only way we could do that is, I guess, if we make the following adjustment, and instead allocating bigger VMA's of size `alloc_size + align` and returning the `vaddr` as `vaddr = ALIGN_UP(vaddr, align)`, but at that point it becomes wasteful I would say as this is a specialized allocator we use in very specific cases, hence why I agreed to the `unlikely`'s as well (I also like them cus they make me look smart). I will add a comment in case this will be needed in the future to know what to do, but I highly doubt it. (yes i am also too lazy to make the change now :cowboy_hat_face: )

File: lib/ukvmem/Config.uk
Comment (michpappas) [reply lib/ukvmem/Config.uk:76]: Random application (e.g. worpress) sounds like an arbitrary criterion though. Similarly to timeouts, there's no such thing as a good default, as there will always either too much or too little. Personally I would choose 1 and let the user adjust to the application requirements. Unless we have strong evidence of an average.

@skuenzer what are your thoughts on this?

File: lib/ukvmem/include/uk/vma_types.h
Comment (michpappas) [reply lib/ukvmem/include/uk/vma_types.h:127]: Yes, but that sounds that they are now incomplete as they only test one guard page?

File: lib/ukvmem/Config.uk
Comment (michpappas) [reply lib/ukvmem/Config.uk]: Okay, but what is the rationale for that? @skuenzer

File: lib/ukallocstack/stack.c
Comment (michpappas) [reply lib/ukallocstack/stack.c:173]: Okay since we have a fundamental limitation on alignment by guards, perhaps we should at least document this constraint, given that it deviates from the expected behavior, and also add a check.

File: lib/ukallocstack/Config.uk
Comment (michpappas) [inline lib/ukallocstack/Config.uk]: "Useful"

File: lib/ukallocstack/include/uk/allocstack.h
Comment (michpappas) [inline lib/ukallocstack/include/uk/allocstack.h]: Nit: "See comment in". 

Also consider whether it would be useful to additionally remind the reader that this is in addition to the usual requirement that alignment must be a power of 2.

File: lib/ukallocstack/include/uk/allocstack.h
Comment (michpappas) [inline lib/ukallocstack/include/uk/allocstack.h]: I don't know why `SOME` initially confused me, `STACK_SIZE` would had been immediately clear to me. But this may be subjective, so if you think it's good as is please feel free to ignore.

File: lib/ukallocstack/stack.c
Comment (michpappas) [reply lib/ukallocstack/stack.c:78]: What I actually meant by `stack_size < initial_size` was under that condition we ask madvise to page in more memory than the actual stack. Therefore we need a check somewhere that `initial_size <= size`.

**Review by michpappas — CHANGES_REQUESTED** [review]
@mogasergiu a few comments on the last update.

File: lib/ukallocstack/include/uk/allocstack.h
Comment (mogasergiu) [reply lib/ukallocstack/include/uk/allocstack.h]: Will surround it with `< >` to further suggest it's just some placeholder.

File: lib/ukschedcoop/include/uk/schedcoop.h
Comment (michpappas) [inline lib/ukschedcoop/include/uk/schedcoop.h:47]: Leaving this as a note for @skuenzer. Not sure if it's something to do in the scope of this PR, but perhaps it would be worth to consider passing a struct with allocators similarly to `ops`. Throughout these PRs I have noticed that many functions in `thread.h` now take 10 or more parameters.

**Review by michpappas — CHANGES_REQUESTED** [review]
From my side it's good. Remaining work before RB:
* Update tests / justify that they don't need to be updated.
* Open questions that require input from @skuenzer.
*  Break down into per-library commits before we merge.

@skuenzer this one's on you now.

File: lib/ukvmem/Config.uk
Comment (skuenzer) [reply lib/ukvmem/Config.uk:76]: In fact the issue was related to on-stack array access. Since stacks are growing downwards and arrays upwards, you easily get outside of a single 4k page when accessing index 0 of a bigger array. On one hand I agree with you @michpappas and on the other hand this should be a protective feature. I found that there is a [POSIX interface](https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_attr_getguardsize.html) for it, so it should be ultimately a per-thread property to make it right. I can imagine that the rational behind this coulb probably be that an application can estimate typical access distances and size the guard pages accordingly.
At this point, I would accept the 4 pages for user experience but longer term we need to change the integration so that it becomes configurable per thread -> I think this requires some API discussion first.

**Comment by mogasergiu** [comment]
TODO: Split last commit into three
- introduce the lib
- extend schedcoop function for the other three allocators
- adjust ukboot to use stack allocator

**Comment by mogasergiu** [comment]
TODO: Following a discussion it's been decided:
- use only 1 guard page for bottom guards
- instead of `initial_len`/`initial_size` use `premapped_len`
- remove the termination function: "Since we generally cannot do it right, do not do it for now"

**Review by michpappas — APPROVED** [review]
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

File: lib/ukschedcoop/include/uk/schedcoop.h
Comment (skuenzer) [reply lib/ukschedcoop/include/uk/schedcoop.h:47]: I understand what you mean. For now, the creation functions for scheduler initialization are custom and defined by the scheduler implementation, likewise the allocator initialization. There is no common init interface layout defined yet for schedulers and allocators. I am also not sure if we could define one because schedulers and allocators could handle very specific cases with very custom parameters.

Comment (skuenzer) [reply lib/ukschedcoop/include/uk/schedcoop.h:47]: Concluding: I think it is okay to expand this definition with this PR. The only place where the scheduler is initialized is in ukboot atm, impact should be minimal for this particular case.

File: lib/ukboot/boot.c
Comment (skuenzer) [inline lib/ukboot/boot.c:68]:
```c
#if CONFIG_LIBUKBOOT_ALLOCSTACK
```
etc. is missing.

Comment (skuenzer) [inline lib/ukboot/boot.c:69]: The following patch makes ukboot still compile even without allocator initialzation:
```diff
--- a/lib/ukboot/boot.c
+++ b/lib/ukboot/boot.c
@@ -65,11 +65,13 @@
 #include <uk/tinyalloc.h>
 #define uk_alloc_init uk_tinyalloc_init
 #endif
+#if CONFIG_LIBUKBOOT_ALLOCSTACK
 #include <uk/allocstack.h>
 #if CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER
 #define ALLOCSTACK_INITIAL_SIZE                                \
        (PAGE_SIZE * (1 << CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER))
 #endif /* CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER */
+#endif /* CONFIG_LIBUKBOOT_ALLOCSTACK */
 #if CONFIG_LIBUKSCHED
 #include <uk/sched.h>
 #endif /* CONFIG_LIBUKSCHED */
```

File: include/uk/plat/config.h
Comment (skuenzer) [inline include/uk/plat/config.h:47]: With re-arch we definitely have to fix the namespacing of the definitions in this file. For now, okay since the newly introduced parameters follows the given examples.

File: lib/ukvmem/include/uk/vma_types.h
Comment (skuenzer) [inline lib/ukvmem/include/uk/vma_types.h]: Could we at least namespace those, like `UK_VMA_STACK_TOP_GUARD_SIZE`? We have previous definitions in this file already namespaced.

File: lib/ukallocstack/include/uk/allocstack.h
Comment (skuenzer) [inline lib/ukallocstack/include/uk/allocstack.h:55]: Can this `maybe_unused` really happen? With `CONFIG_LIBUKVMEM` on, it should be used always, right? I would also prefer if we add this attribute to the final function declaration instead.

**Review by skuenzer — APPROVED** [review]
Thanks a lot for this work!

Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request merged**
